### ISSUE TITLE
Make fragments usable

### ISSUE THREAD

**Issue by pekam** [vaadin/flow#3588]
TLDR: The way that our bundling currently works makes it impossible to create useful fragments. The shell-file often depends on the files you're trying to separate to a fragment for two reasons:
1. We add all the `HtmlImports` to shell-file, even if we don't use them.
2. If a file is referenced from inside any template, it's a dependency of the shell-file and can't be separated.

---
Currently we scan for all the `HtmlImports` in the classpath and create a shell-file (vaadin-flow-bundle.html) out of those, to have one huge bundle with all the frontend files. This means that if you're depending on the platform, you get imports for all the Vaadin elements.

When we define fragments, we are creating new entrypoints to new bundles (fragments) with links/imports to frontend-files just like in the shell-file. To define which sources to include in which bundle, polymer-bundler travels through all the dependencies of the entrypoints and includes those in the same bundle. When there are shared dependencies between two bundles, another bundle-file (called `shared_bundle1.html` or something) is created for those. This logic is located [here](https://github.com/Polymer/polymer-bundler/blob/e573200de6a3782e288b99149db5aeec3c1bd27f/src/bundler.ts#L156). After doing this dependency analysis, a merge-stategy is applied that handles the shared_bundles. The default merge strategy just combines all the shared_bundles in the main-bundle.

What this means in practice is that we can't add to a fragment anything that is a transitive dependency of any `HtmlImport`.
Say that you have only one view using `vaadin-button` and `vaadin-checkbox`, so you decide to create a fragment for those, so that they're not included in the main-bundle and requested only when the user visits that page. The maven plugin knows to remove those imports from the shell-file when they are included in a fragment, so the entrypoints don't share direct dependencies. If your app is depending on the platform, this doesn't work though, because `vaadin-grid` depends on `vaadin-checkbox` and `vaadin-upload` depends on `vaadin-button`. Even if you never use grid or upload elements, those are included in the shell-file, and button and checkbox become shared dependencies between the main-bundle and the fragment, so they are merged back to the main-bundle.

Similarly, I can't separate one template-view with it's unique dependencies into it's own fragment if the template is referenced from another template. I tried to make separate fragment for Bakery's login-view, but since [shell-file depends on it through main-view](https://github.com/vaadin/bakery-app-starter-flow-spring/blob/81fbe5655d22aa49de8985d07da416372eb5d5c3/src/main/webapp/frontend/src/main-view.html#L7), it's added back to the main-bundle as a shared dependency.

**Comment by manolo** [comment]
IMO, there are two sceneries
- Pure java UI. The best option is that WC in classpath are automatically included in the bundle.
- UI made with templates. If the user takes care of adding all imports correctly in all views like in Bakery, there is no need for looking for components in java classes. In the case where there is some component created via java API, the user always can import that in some template or in the main view.

Whether to look for components in the classpath might be configured via a flag in maven plugin, something like `<autoDiscoverComponents>false<autoDiscoverComponents>` or disabled if there are `<fragments>`

To me, having auto-discover enabled when trying to optimise fragments have more caveats than advantages.

**Comment by pleku** [comment]
As the html imports are now deprecated, it makes no sense to improve the bundling for those.

For npm mode, there is #5537
